Make cypress psoc64 TFM ready and also add TF-M initialization

jainvikas8 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Non-secure core IPC, Mailbox and Interface init for TFM_DUALCPU and TFM_V8M targets

These changes concentrate on getting mbed-os(Non-secure) working with TF-M(secure)
Changes reflect towards any Cypress based TFM_DUALCPU targets and TFM_V8M targets at the moment and making it more generic in the near future.

Key changes:
* Init IPC, Mailbox and Non-secure Interface.
   These changes are for TFM_DUALCPU and TFM_V8M platforms.
   The initialization happens after mbed-os kernel is kicked off and
   when the thread is up and running.

   We are initializing the following for TFM_DUALCPU platforms:
   * IPC Interrupts for syncing multi-core platforms.
   * NS Mailbox to receive messages.
   * NS interface.

   We are only initializing a Non-secure interface for TFM_V8M platforms.

* Cypress: CY8CKIT_064S2_4343W: Reserve timer channels used by TF-M.
* Make CY8CKIT_064S2_4343W Cypress target TF-M compatible.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->

This impacts CY8CKIT_064S2_4343W, TFM_V8M and TFM_DUALCPU targets.
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behavior change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

These changes are tested on 
* CY8CKIT_064S2_4343W with M0+ running TF-M Secure binary and M4 running blinky example with mbed-os non-secure binary.
* ARM_MUSCA_A1 with TFM + mbed-os binary running blinky example.

The blinky example is tested with:
```mbed compile -t ARMC6 -m CY8CKIT_064S2_4343W ```
```mbed compile -t GCC_ARM -m ARM_MUSCA_A1 ```

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

@Patater @devran01 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

urutva commented on platform/mbed_lib.json:
```diff
@@ -159,6 +159,10 @@
         },
+        "CY8CPROTO_064_SB": {
+            "stdio-baud-rate": 115200,
+            "default-serial-baud-rate": 115200
```
Why do we need to set `default-serial-baud-rate`? Aren't we only using DAPLINK serial which should be mapped to `stdio`? If yes, then setting `stdio-baud-rate` should suffice right?

jainvikas8 replied:
I was advised to keep it same by @ashok-rao

urutva commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -32,6 +32,13 @@
 #endif
 
+#if defined(TARGET_TFM_TWINCPU)
+#include "FEATURE_PSA/TARGET_TFM/TARGET_TFM_TWINCPU/inc/tfm_multi_core_api.h"
```
The build tool will automatically add the include path `FEATURE_PSA/TARGET_TFM/TARGET_TFM_TWINCPU/inc/`. It isn't a good idea to hardcode the include path. Same applies to other headers as well.

urutva commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -64,6 +71,45 @@ void mbed_rtos_init()
+                                   MBED_ERROR_CODE_INITIALIZATION_FAILED),
+                   "Failed to sync-up multi-core");
+        /* Avoid undefined behavior after multi-core sync-up failed */
+        for (;;) {
```
This code will be non-reachable as `MBED_ERROR` never returns.

Patater commented on platform/mbed_lib.json:
```diff
@@ -159,6 +159,10 @@
+        "CY8CPROTO_064_SB": {
+            "stdio-baud-rate": 115200,
+            "default-serial-baud-rate": 115200
+        },
```
Would be better to not add this if the serial port works at the default baud. We don't want to surprise users with different behavior compared with other targets.

Patater commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -32,6 +32,13 @@
+#include "FEATURE_PSA/TARGET_TFM/TARGET_TFM_TWINCPU/inc/tfm_multi_core_api.h"
+#include "FEATURE_PSA/TARGET_TFM/TARGET_TFM_TWINCPU/inc/tfm_ns_mailbox.h"
+#include "FEATURE_PSA/TARGET_TFM/TARGET_TFM_TWINCPU/inc/platform_multicore.h"
+#include "FEATURE_PSA/TARGET_TFM/include/tfm_ns_interface.h"
```
Are all these headers really needed? Are any unused?

jainvikas8 replied:
Yes. will remove unused.

urutva commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -64,6 +71,45 @@ void mbed_rtos_init()
+                                   MBED_ERROR_CODE_INITIALIZATION_FAILED),
+                   "Failed to initialize NS mailbox");
+        /* Avoid undefined behavior after NS mailbox initialization failed */
+        for (;;) {
```
This code will be non-reachable as `MBED_ERROR` never returns.

urutva commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -64,6 +71,45 @@ void mbed_rtos_init()
+                                   MBED_ERROR_CODE_INITIALIZATION_FAILED),
+                   "Failed to initialize NS interface");
+        /* Avoid undefined behavior after NS interface initialization failed */
+        for (;;) {
```
This code will be non-reachable as `MBED_ERROR` never returns.

Patater commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -64,6 +71,45 @@ void mbed_rtos_init()
+                   "Failed to sync-up multi-core");
+        /* Avoid undefined behavior after multi-core sync-up failed */
+        for (;;) {
+        }
```
Do we really need the infinite for loop here? `MBED_ERROR()` will call `mbed_halt_system()`, which should either loop forever efficiently (including a `WFE` in the loop, so the core sleeps and doesn't waste energy) or call `mbed_die()`, which will sleep and blink an error pattern forever.

If you think you really need the loop here, we have a bug in `MBED_ERROR()` on this platform as it isn't halting the system properly. We should fix that bug instead of adding the infinite loop here.

Patater commented on rtos/source/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -107,6 +153,10 @@ MBED_NORETURN void mbed_rtos_start()
 #endif // defined(FEATURE_TFM) && defined(FEATURE_NSPE)
 
+#if defined(TARGET_TFM_TWINCPU)
+    tfm_ns_multi_core_boot();
```
I am not sure if `mbed_rtos_start()` is the correct place to hook in behavior that applies to all `TFM_TWINCPU` targets. Do we have a better place to hook? A worse place to hook would be in target specific code that would require copy-pasting for any new `TFM_TWINCPU` target; that'd be a bad experience for folks bringing new TF-M targets to Mbed OS.

@bulislaw Any recommendations? Is this really the best place?

bulislaw replied:
I'm not sure what you need/want to achieve and by when. Actually adding a comment for this function's intent could be good. Don't get me wrong I can read what it does, but I lack background to actually understand implications and the best place to do it or the requirements. Looking at the other changes it has dependency on RTOS - mutex, so that would require RTOS to actually run. The bootstrap process is described here https://github.com/ARMmbed/mbed-os/blob/master/rtos/source/TARGET_CORTEX/mbed_boot.h

Patater commented on targets/TARGET_Cypress/TARGET_PSOC6/mbed_rtx.h:
```diff
@@ -21,6 +21,10 @@
+ * This is used to obtain Object specific Memory allocation required for
+ * mutex creation in an efficient way.
+ */
+#define OS_MUTEX_OBJ_MEM    1
```
Is this the best place to add this? This seems like code any TFM_TWINCPU target would need to add, and not PSoC 6 specific. Can we give folks adding new TF-M targets to Mbed OS an easier time of things (i.e. require less copy paste code of them)?

As more acceptable option, we could add an `#ifdef` in `rtos/source/TARGET_CORTEX/mbed_rtx_conf.h` to configure the necessary memory pools (probably eventually more than just mutexes) for all `TFM_TWINCPU` at once, making porting new targets easier.

jainvikas8 replied:
This is more of architectural decision, rather not alter - ```rtos/source/TARGET_CORTEX/mbed_rtx_conf.h``` , but need to find a commonplace that could influence the right platform if the above is not suitable.

Patater replied:
I don't think adding this configuration for each target is acceptable from a "make adding new hardware easier" perspective. We need to find a shared place to inject RTX configuration from.

jainvikas8 replied:
Just to be clear, In this case, it adds to all Cypress platform.

bulislaw commented on targets/TARGET_Cypress/TARGET_PSOC6/mbed_rtx.h:
```diff
@@ -21,6 +21,10 @@
+ * This is used to obtain Object specific Memory allocation required for
+ * mutex creation in an efficient way.
+ */
+#define OS_MUTEX_OBJ_MEM    1
```
Sorry guys, this is not supported memory model in Mbed OS. Please use statically allocated memory for your mutexes.

bulislaw reviewed (dismissed):
RTX object memory is currently not supported by Mbed OS, we opted out not to mix different memory models in the same system. The object specific pool is not very good for generic use case as we can't plan in advance how many objects will be needed, leading to user confusion and/or wasting memory. Please find a way of using statically allocated memory for the mutexes.

Patater committed:
cypress: psoc64: Add TF-M compatibility

Make the CY8CKIT_064S2_4343W target TF-M compatible by addding flash and
region definitions from TF-M (at c4f37c18c4a0) and by updating the
CY8CKIT_064S2_4343W linker script to create a flash image compatible
with TF-M.

Signed-off-by: Jaeden Amero <jaeden.amero@arm.com>

Patater commented on platform/mbed_lib.json:
```diff
@@ -159,6 +159,9 @@
         },
+        "CY8CPROTO_064_SB": {
+            "stdio-baud-rate": 115200
+        },
```
Please remove commit "Change CY8CPROTO_064_SB target serial baud." from this PR

jainvikas8 replied:
I think this is fairly important, as TF-M secure image runs on 115200 baud and we don't want to switch the serial when the mbed-os switches to 9600.

Patater replied:
Users of Mbed OS expect to be able to use 9600 baud by default, and this breaks Mbed OS user expectations. Perhaps we can change Mbed OS's TF-M building script to set the TF-M baud to 9600 instead.

jainvikas8 replied:
I think this is the safest place to change the baud rate, as there are other targets listed here with similar requirements. We are still using mbed-cli to compile the Non-secure image, which TF-M building script are we talking about?

Patater replied:
The other targets are not good examples of best practice. Best practice is not to surprise the user.

The TF-M building script we are talking about is the one that builds the secure image: `tools/psa/build_tfm.py`. I'm suggesting changing the baud used in the TF-M binaries we ship with Mbed OS to match what Mbed OS users expect.

jainvikas8 replied:
If we are talking about changing the baud rate in TF-M secure binary then we need to upload a new patch to https://git.trustedfirmware.org/trusted-firmware-m.git/tree/platform/ext/target/cypress/psoc64/Device/Config/device_cfg.h. I think there is no way we could change the baud rate during compile time to 9600 to match mbed-os binary

Patater replied:
We can change this in our TF-M build script later, after TF-M becomes more easily configured.

Patater commented on features/FEATURE_PSA/TARGET_TFM/TARGET_TFM_DUALCPU/src/tfm_mbed_boot.c:
```diff
@@ -0,0 +1,58 @@
+/* mbed Microcontroller Library
```
In commit message "Init IPC, Mailbox and Non-secure Interface. ", you refer to "TFM_TWINCPU", but you renamed it to "TFM_DUALCPU" in the previous commit already.

jainvikas8 replied:
Fixed

Patater commented on features/FEATURE_PSA/TARGET_TFM/TARGET_TFM_DUALCPU/src/tfm_mbed_boot.c:
```diff
@@ -0,0 +1,58 @@
+void mbed_main(void)
+{
+    /*
+     * In case the of Dual Core, we need to initialize the IPC, Mailbox
```
```diff
- Dual Core
+ dual CPU
```

jainvikas8 replied:
Fixed.

Patater commented on features/FEATURE_PSA/TARGET_TFM/TARGET_TFM_DUALCPU/src/tfm_mbed_boot.c:
```diff
@@ -0,0 +1,58 @@
+void mbed_main(void)
+{
+    /*
+     * In case the of Dual Core, we need to initialize the IPC, Mailbox
```
```diff
- the IPC, Mailbox
+ IPC and the mailbox
```

jainvikas8 replied:
Fixed.

Patater commented on features/FEATURE_PSA/TARGET_TFM/TARGET_TFM_DUALCPU/src/tfm_mbed_boot.c:
```diff
@@ -0,0 +1,58 @@
+    int32_t ret;
+
+    ret = tfm_ns_wait_for_s_cpu_ready();
+    if (ret != PLATFORM_MAILBOX_SUCCESS) {
```
I would have expected "TFM_SUCCESS" to be used here by TF-M, as this isn't a mailbox function. It just so happens that on PSoC 6 platforms the implementation uses a mailbox, but other platforms may not. The PSoC 6 implementation of tfm_platform_ns_wait_for_s_cpu_ready() is returning too low level of an error code and should instead return `TFM_SUCCESS`.

jainvikas8 replied:
This would require changes to TF-M as these files are imported as part of the integration. I could remove `PLATFORM_MAILBOX_SUCCESS` and check as `if(!ret)` instead?

Patater replied:
Yes, this requires changes to TF-M. The bug is in TF-M code.

jainvikas8 replied:
On closer investigation - https://git.trustedfirmware.org/trusted-firmware-m.git/tree/app/main_ns.c#n90, seems like they never check the return type.

Have created - https://developer.trustedfirmware.org/T660

Should I remove the return type check and keep it simple as TF-M implementation for time being?

Patater replied:
It'd be quicker to fix TF-M directly.

Patater replied:
Could you please link to the TF-M issue, and explain what's wrong with the code as currently written, in a code comment here? We should ensure we don't keep this bad code here forever. Linking to the issue and noting that we can use TFM_SUCCESS after it is fixed will help us know when we can fix the code and why the code is currently not quite right.

Patater commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017-2019, Arm Limited. All rights reserved.
+ * Copyright (c) 2017-2020, Arm Limited. All rights reserved.
```
In commit messge "Change OS Wrapper layer", the title "Change OS Wrapper layer" isn't very precise. Consider saying something like "Allocate TF-M objects from the heap" instead.

jainvikas8 replied:
Fixed.

Patater commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c:
```diff
@@ -11,31 +11,59 @@
-    if (stack_size != OS_WRAPPER_DEFAULT_STACK_SIZE) {
-        task_attribs.stack_size = stack_size;
+    thread = (osRtxThread_t *)malloc(osRtxThreadCbSize);
+    if(thread == NULL){
```
Does this code pass AStyle? Is AStyle being run on this code?

Both Mbed OS and TF-M style is to have a space after keywords like `if`.

jainvikas8 replied:
This file is in the ignore list --  `.astyleignore`
`features/FEATURE_PSA/TARGET_TFM`
Since we are going to maintain it, I can format it, but it can't be caught by the CI, because the whole folder is ignored.
So let me know your thoughts.

jainvikas8 replied:
Fixed.

Patater replied:
We should use Mbed OS style and change the .astyleignore to ensure we have CI checks on style.

jainvikas8 replied:
I'm not sure how this can be achieved
The importer uses this json - `tools/psa/tfm_ns_import.json` to not only import individual files but folders too, therefore **.astyleignore** completely ignores `features/FEATURE_PSA/TARGET_TFM`
So, is there a mechanism in mbed-os to check a particular file for astyle where a complete directory is ignored for check?

Patater replied:
Modify .astyleignore to not ignore the entire folder. Add back all folders and files you do want ignored.

jainvikas8 replied:
Fixed

Patater commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c:
```diff
@@ -11,31 +11,59 @@
+        .cb_mem = thread,
+        .cb_size = osRtxThreadCbSize,
+        .stack_size = stack_size,
+        .stack_mem = stack};
```
Could we use Mbed OS C++ Thread class instead? Then we don't need to duplicate the know-how of how to create an Mbed OS thread here. For instance, if the Mbed OS C++ Thread class decided to change if threads were "joinable", we wouldn't pick up that change here, even though in current implementation the behavior is aligned. This wrapper code would be more robust if we'd avoid repeating ourselves (DRY) and use the C++ Thread class.

jainvikas8 replied:
Mbed OS C++ thread class always sets the thread to be detached state by default and have not seen any occurrences of joinable state using mbed-os C++ class. I don't see a reasonable argument to make so many changes (almost rewriting the wrapper) to increase the robustness at the moment.

Patater replied:
Joinable state was just one example. Any part of this could change. For instance, the memory layout of mutexes in RTX could change, requiring updates to both here and the Mutex class, rather than just the Mutex class.

Maintainability is an extremely strong reason to block adding this code in its current shape.

jainvikas8 replied:
I think we need to align with TF-M, as these wrappers are only to be used with them. If Mbed OS C++ decides to make it as "joinable" and if TF-M API's think them to be "Detachable" thread then it might be a problem.

If my understanding is correct, For example, the regression tests don't use `thread_join()` API, they just support mostly "detachable" state operations.

Patater replied:
Anything could change, not just joinability. For instance, Mbed OS may want to change how stacks are allocated. Having one place for the logic to allocate stacks is best for maintainability. Let's re-use the Mbed OS code we have rather than reimplementing it again in C and just for TF-M; it's better for maintainability and also for code size (since we'll have less code).

Patater commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c:
```diff
@@ -11,31 +11,59 @@
+        .attr_bits = 0,
+        .cb_mem = semaphore,
+        .cb_size = osRtxSemaphoreCbSize
+    };
```
Consider using the Mbed OS C++ Semphore class instead.

jainvikas8 replied:
Fixed in a later commit.

Patater commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c:
```diff
@@ -75,21 +103,25 @@ uint32_t os_wrapper_semaphore_delete(void *handle)
-        .cb_size = 0U
+        .attr_bits = osMutexRecursive | osMutexPrioInherit | osMutexRobust,
+        .cb_mem = mutex,
+        .cb_size = osRtxMutexCbSize
```
Consider using the Mbed OS C++ Mutex class instead.

jainvikas8 replied:
Fixed in a later commit.

Patater commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c:
```diff
@@ -168,3 +202,41 @@ void os_wrapper_thread_exit(void)
+    }
+
+    free(st_mem);
+    free(handle);
```
Call `delete` on the Mbed OS C++ Thread object instead. Use the address of the thread object as your `void *handle`.

jainvikas8 replied:
Fixed in a later commit.

Patater commented on tools/psa/tfm_ns_import.json:
```diff
@@ -28,10 +28,6 @@
-        {
-            "src_file": "app/os_wrapper_cmsis_rtos_v2.c",
-            "dst_file": "features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c"
-        },
```
"Remove TF-M import of os_wrapper_cmsis_rtos_v2.c" should be done before or as part of the commit "Change OS Wrapper layer". This is a better conceptual organization because we could run the importer at every commit and expect no bugs added from the import list and our own modifications being out of sync. Importing at commit "Change OS Wrapper layer" in current organization would break by overwriting our changes.

jainvikas8 replied:
Fixed.

Patater commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CPROTO_064_SB/device/COMPONENT_CM4/system_psoc6_cm4.c:
```diff
@@ -171,7 +171,7 @@ void SystemInit(void)
 
 #ifdef __CM0P_PRESENT
-    #if (__CM0P_PRESENT == 0)
+    #if (__CM0P_PRESENT == 1)
```
Don't we need to also update this for `TARGET_CY8CKIT_064S2_4343W`? We should also have Cypress review this change.

jainvikas8 replied:
Yes, cypress needs to review for `CY8CPROTO_064_SB`. But I'm unsure if this update is required for `TARGET_CY8CKIT_064S2_4343W`, @devran01 can you comment on this, please?

urutva replied:
@Patater @jainvikas8 We don't need this change in `TARGET_CY8CKIT_064S2_4343W`, as the mailbox semaphores will be initialised when mailbox is initialised rather than `systeminit()`.

Patater replied:
Why couldn't the mailbox semaphores for `TARGET_CY8CPROTO_064_SB` be initialized when the mailbox is initialized? Seems odd that `TARGET_CY8CPROTO_064_SB` or `TARGET_CY8CKIT_064S2_4343W` would be different in this regard.

urutva replied:
@jainvikas8 Why is this commit still present in the PR. Shouldn't this go away since the mailbox semaphores will be initialised when mailbox is initialised?

jainvikas8 replied:
@devran01 This change is required for `TARGET_CY8CPROTO_064_SB`, mailbox semaphores will not be initialized. Without this, the mbed-os blinky application fails.

urutva replied:
@jainvikas8 The patch https://github.com/Patater/trusted-firmware-m/commit/02a132b30992f25b9ea6d649250b35a3718d0f0d initialises mailbox semaphores during mailbox init. Therefore we don't need to have this commit.

jainvikas8 replied:
Fixed.

urutva committed:
cypress: psoc64: Reserve timer channels used by TF-M

There are two timers, Timer0 and Timer1, available on the PSoC64. Timer0
has 8 channels and Timer1 has 24 channels. TF-M regression tests make
use of Timer0 Channel 0 and Timer0 Channel 1. Therefore, reserve the
timer channels used by TF-M. This approach can be replaced once we have
a way to allocate dedicated timers for TF-M and Mbed OS.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>
Signed-off-by: Vikas Katariya <vikas.katariya@arm.com>

urutva commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.c:
```diff
@@ -11,31 +11,60 @@
-    task_attribs.priority = (osPriority_t) priority;
 
-    return (void *)osThreadNew(func, arg, &task_attribs);
+    stack = (void *)malloc(stack_size);
```
why is the check for `stack_size != OS_WRAPPER_DEFAULT_STACK_SIZE` removed? The value of `OS_WRAPPER_DEFAULT_STACK_SIZE` is -1. So if tests wants to use default stack size then -1 will be passed as argument. `malloc` will fail in that case.

jainvikas8 replied:
When we are managing the memory manually, we don't have a default size of the stack which can be allocated by the OS. There it would be good to force the developer to specify the size required.

urutva replied:
> Patch - [84c3916](https://github.com/ARMmbed/mbed-os/commit/84c39161ca39fbf8298032038b1f05ea50f41a7d) is going to be upgraded with new Mbed OS C++ implementation as requested by the reviewer. On a standalone basis, this commit works, should not cause any harm.
> 
> Patch - [d1ff232](https://github.com/ARMmbed/mbed-os/commit/d1ff232a5d389badcd55df409fc45901b0aeeb91) is required for `TARGET_CY8CPROTO_064_SB` since this a prototype and development is been halted. Seeing this target has played an important role in bring-up of TF-M + Mbed-os integration, therefore this patch has been added.

@jainvikas8 The changes done in the commit https://github.com/ARMmbed/mbed-os/tree/ad7014054b7974b48a63bef2c2066b93714da2e3 are removed by https://github.com/ARMmbed/mbed-os/tree/daf9fddbb2fb7f81e3944a2ec26684014db90b47. So why do we want to keep previous one?

jainvikas8 replied:
Because this is a design change, On a standalone basis this commit will ok.

jainvikas8 reviewed (commented):
Patch - https://github.com/ARMmbed/mbed-os/commit/84c39161ca39fbf8298032038b1f05ea50f41a7d is going to be upgraded with new Mbed OS C++ implementation as requested by the reviewer. On a standalone basis, this commit works, should not cause any harm.

Patch - https://github.com/ARMmbed/mbed-os/commit/d1ff232a5d389badcd55df409fc45901b0aeeb91 is required for `TARGET_CY8CPROTO_064_SB` since this a prototype and development is been halted. Seeing this target has played an important role in bring-up of TF-M + Mbed-os integration, therefore this patch has been added.

urutva commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.cpp:
```diff
@@ -0,0 +1,168 @@
+
+extern "C" uint32_t os_wrapper_semaphore_delete(void *handle)
+{
+    delete (rtos::Semaphore *)handle;
```
Make sure that semaphore is released before freeing allocated memory.

Patater replied:
It should be up to callers of `os_wrapper_semaphore_delete()` to ensure that's the case.

urutva replied:
@Patater I do not agree. It is the responsibility of the OS wrappers to make sure that semaphore is in released state before deleting allocated memory. Imagine a scenario where a semaphore was allocated, locked and then deleted. If the same memory was allocated to another semaphore (even though low possibility still a possibility) then that semaphore will be in locked state forever.

Patater replied:
Can we depend on the Mbed OS C++ Semaphore class to do this checking for us? We shouldn't duplicate checks that it would do.

jainvikas8 replied:
According to this - https://github.com/ARMmbed/mbed-os/blob/feature_twincpu/rtos/source/TARGET_CORTEX/rtx5/RTX/Source/rtx_semaphore.c#L331
This unblocks the waiting threads which are waiting for it.

Patater replied:
As Mbed OS doesn't check this, so we shouldn't either.

urutva commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.cpp:
```diff
@@ -0,0 +1,168 @@
+
+extern "C" uint32_t os_wrapper_mutex_delete(void *handle)
+{
+    delete (rtos::Mutex *)handle;
```
Make sure that mutex is released before freeing allocated memory.

jainvikas8 replied:
I think the same comment applies here - https://github.com/ARMmbed/mbed-os/pull/12271#discussion_r375267774

jainvikas8 replied:
According to this - https://github.com/ARMmbed/mbed-os/blob/feature_twincpu/rtos/source/TARGET_CORTEX/rtx5/RTX/Source/rtx_mutex.c#L376
should not be problem, if delete is called.

urutva commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.cpp:
```diff
@@ -0,0 +1,168 @@
+    /* This enables thread context management for trustzone */
+    uint32_t tz_module = 1;
+
+    new_thread = new rtos::Thread(tz_module, (osPriority_t) priority, stack_size, NULL, name);
```
TF-M implementation of the wrapper has check for stack_size == -1 for default stack size, therefore it might be possible that some of the tests are passing the stack_size as -1. Therefore it is necessary to check the stack_size before passing it to Thread constructor.

jainvikas8 replied:
I can check the `stack_size` to -1 but if it is then can't allocate anything as there is NO default size set, would be ok to return? At the moment if -1 is used then class doesn't have default size - https://github.com/ARMmbed/mbed-os/blob/feature_twincpu/rtos/source/Thread.cpp#L45

Patater replied:
We could use the Mbed OS C++ Thread class's default stack size, when -1 is passed by wrapper users, by not specifying the stack size to create in the constructor. This would provide an implementation of the wrappers that does what wrapper users expect.

Patater replied:
Oops. Seems I'm imagining constructors that don't exist. Anyhow, there ought to be an Mbed default stack size macro we can use.

Patater replied:
Possibly `OS_STACK_SIZE` as defined by `rtos/source/TARGET_CORTEX/mbed_rtx_conf.h`, which reads from the Mbed configuration system (i.e. all that json stuff).

Patater replied:
Actually yes, there is such a constructor. We should use it. 

Before, I was reading the cpp instead of the header file.

See https://github.com/ARMmbed/mbed-os/blob/feature_twincpu/rtos/Thread.h#L100-L105

jainvikas8 replied:
Yes, that can be done - please note - `OS_STACK_SIZE` is **4096 bytes**. Fixed.

urutva commented on features/FEATURE_PSA/TARGET_TFM/src/os_wrapper_cmsis_rtos_v2.cpp:
```diff
@@ -0,0 +1,168 @@
+
+    new_thread = new rtos::Thread(tz_module, (osPriority_t) priority, stack_size, NULL, name);
+
+    new_thread->start(mbed::callback(func, arg));
```
Make sure `new_thread` is not NULL before invoking `start` method on that object.

jainvikas8 replied:
fixed.

urutva commented on rtos/source/ThisThread.cpp:
```diff
@@ -251,4 +251,32 @@ const char *get_name()
+
+    return osThreadGetPriority(id);
+#else
+    return (osPriorityError)(-1);
```
The definition of `osPriorityError` might not be available if `MBED_CONF_RTOS_PRESENT ` is not present.

jainvikas8 replied:
Casting should work because functions are returning the same type. Example: https://github.com/ARMmbed/mbed-os/blob/master/rtos/source/ThisThread.cpp#L232

urutva commented on rtos/source/ThisThread.cpp:
```diff
@@ -251,4 +251,32 @@ const char *get_name()
+
+    return osThreadSuspend(id);
+#else
+    return (osError)(-1);
```
The definition of `osError ` might not be available if `MBED_CONF_RTOS_PRESENT ` is not present.

jainvikas8 replied:
Casting should work because functions are returning the same type. Example: https://github.com/ARMmbed/mbed-os/blob/master/rtos/source/ThisThread.cpp#L232

Patater replied:
Can we try a baremetal build and see what breaks? Will CI already try for us?

jainvikas8 replied:
Did try that, it doesn't compile ThisThread.cpp, Semaphore.cpp or Mutex.cpp
https://github.com/armmbed/mbed-os-example-blinky-baremetal/

Patater replied:
You might need to attempt to use a Semaphore in order to properly test. Otherwise, what would be the point of the ifdefs in those classes... There must be a way to get those classes included.

jainvikas8 replied:
yes, tried a different method, the way to include is changing `"requires": ["bare-metal","rtos-api"]` in mbed_app.json on https://github.com/armmbed/mbed-os-example-blinky-baremetal/
Fixes to follow soon.

Patater commented on rtos/source/ThisThread.cpp:
```diff
@@ -251,4 +251,32 @@ const char *get_name()
+#else
+    return static_cast<osStatus_t>(osError);
+#endif
+}
```
The changes to ThisThread and rtos/mbed_rtos_types.h should be done in their own commit with their justification clearly spelled out in the commit message description.

jainvikas8 replied:
If you referring to splitting bare-metal changes? then the bare-metal build will break because of the commit - https://github.com/ARMmbed/mbed-os/pull/12271/commits/be6f12bb19f3be1513c7ff94480811052e9a46af.

The OS wrapper which utilizes the `suspend and get_priority` methods needs to be defined in the single commit both for bare-metal and RTOS builds.

Patater replied:
Do the changes in a commit on their own before any OS wrapper commits that require the changes. You are adding new functionality to Mbed OS C++ RTOS API. These could be upstreamed to master today.

I don't propose to separate baremetal and RTOS changes, but to do both baremetal and RTOS changes together separate from any TF-M OS wrapper changes.

jainvikas8 replied:
If I just split "Mbed OS C++ class" and "TF-M OS-Wrapper" changes separately, then "OS-Wrapper changes won't work on their own, as they have a dependency on the "Mbed OS C++ class" changes.
I understand the sequence you are mentioning
1st -> Mbed OS C++ Class changes
2nd -> TF-M Wrapper changes, but it would be nice to avoid this if any reordering mishaps happen.

Patater replied:
Yup. You should also raise a separate PR to master for the C++ changes. I think @kjbracey-arm should review the changes for any potential issues.

Patater commented on rtos/mbed_rtos_types.h:
```diff
@@ -58,6 +58,62 @@ typedef enum {
+    osPriorityRealtime6     = 48 + 6,     ///< Priority: realtime + 6
+    osPriorityRealtime7     = 48 + 7,     ///< Priority: realtime + 7
+    osPriorityISR           = 56,         ///< Reserved for ISR deferred thread.
+    osPriorityError         = -1,         ///< System cannot determine priority or illegal priority.
```
Do we really need all these other priorities? Or would just osPriorityError be fine?

jainvikas8 replied:
This is been taken from cmis_os2.h to match what we are talking about. `osStatus_t` is copied from there. Although we may not require for non-rtos

Patater commented on rtos/source/ThisThread.cpp:
```diff
@@ -251,4 +251,32 @@ const char *get_name()
+
+    return static_cast<osPriority_t>(osThreadGetPriority(id));
+#else
+    return static_cast<osPriority_t>(osPriorityError);
```
We might also want to say osPriorityNormal for baremetal Mbed OS, rather than an error.

jainvikas8 replied:
`osPriorityNormal` refers to 24, are we to use something else?

Patater replied:
osPriorityNormal is the default priority. It makes sense as a sane default to return in baremetal, at least to me.

Patater commented on rtos/source/ThisThread.cpp:
```diff
@@ -251,4 +251,32 @@ const char *get_name()
+
+    return static_cast<osStatus_t>(osThreadSuspend(id));
+#else
+    return static_cast<osStatus_t>(osError);
```
Suspend on baremetal might be better off being a no-op, rather than an error.

The RTOS API for baremetal is about application portability, so folks can run their existing apps. If we make their existing apps start getting errors where things worked before, we unfortunately break portability.

jainvikas8 replied:
I could pass `osOK` = 0, if that would suffice?

Patater replied:
Let's improve these in https://github.com/ARMmbed/mbed-os/pull/12388

I'll accept the code as-is into the feature branch.

Patater reviewed (dismissed):
Looks good for feature branch

Patater reviewed (dismissed):
LGTM

0xc0170 commented on rtos/source/TARGET_CORTEX/mbed_boot.h:
```diff
@@ -149,6 +150,20 @@ void mbed_toolchain_init(void);
+ *
+ * This is a weak function which can be overridden by an application
+ * to allow code to run before mbed_main is called.
+ * Some TF-M supported platforms require syncronisation between
```
`synchronisation`

0xc0170 reviewed (dismissed):
just one typo

jainvikas8 committed:
psa: Init IPC, Mailbox and Non-secure Interface

These changes are for TFM_DUALCPU and TFM_V8M platforms.
The initialization happens after mbed-os kernel is kicked off and
when the thread is up and running.

We are initializing the following for TFM_DUALCPU platforms:
* IPC Interrupts for syncing multi-core platforms.
* NS Mailbox to receive messages.
* NS interface.

We are only initializing Non-secure interface for TFM_V8M platforms.

mbed_tfm_init() in tfm_mbed_boot.c overrides the WEAK mbed_tfm_init()
for a specific platform.

Signed-off-by: Vikas Katariya <vikas.katariya@arm.com>

morser499 commented on targets/TARGET_Cypress/TARGET_PSOC6/cy_us_ticker_api.c:
```diff
@@ -54,6 +54,23 @@ static void cy_us_ticker_irq_handler(MBED_UNUSED void *arg, MBED_UNUSED cyhal_ti
 {
     if (!cy_us_ticker_initialized) {
+
+#ifdef TARGET_TFM
```
I don't understand this change. I don't see anything in this MR that actually uses these. If the timers are only used in regression tests, then they should be reserved by the regression tests themselves, not globally. Also, It says it uses channels 1 & 2, but then reserves 0 & 1.

Patater replied:
We have to prevent Mbed OS from using the timer the regression test application would use. Any suggestions on how to do this from an Mbed OS application, rather than Mbed OS itself, would be much appreciated! Thanks.

morser499 replied:
Do you have the application code that I can take a look at? I expect you don't want to put any Cypress specific code there, but I don't see any of this help without doing so. The reserved times are just at the Cypress layer. Nothing is actually configuring them here.

morser499 commented on targets/TARGET_Cypress/TARGET_PSOC6/cy_us_ticker_api.c:
```diff
@@ -54,6 +54,23 @@ static void cy_us_ticker_irq_handler(MBED_UNUSED void *arg, MBED_UNUSED cyhal_ti
 {
     if (!cy_us_ticker_initialized) {
+
+#ifdef TARGET_TFM
```
This does not make sense to me. If these are only used for testing, why aren't they being reserved by the test itself. That keeps the code together and avoid an unnecessary reservation in the general code.  Also, it says it uses channels 1 & 2, but then reserves 0 & 1.

0xc0170 replied:
My understanding here: us ticker is trying to allocate 2 free timers . How it can work: Alocate 0 - failed, try next one, failed, next one - until there is a free one. Considering all channels are capable of what us ticker requires.
`TARGET_TFM` should not be necessary here, as any app can allocate timer before this runs. 

@morser499 What would you suggest here ?

Patater replied:
@0xc0170 How can an app allocate a timer before this runs?

morser499 replied:
Unfortunatly it looks like I posted twice...
@0xc0170 cyhal_timer_init() will already allocate the first available one. The calls to cyhal_hwmgr_reserve() just prevent those resources from being used.
@Patater I am not sure how the application is actually using these. Can you point me to the application in question to review?

Patater replied:
It uses the hardware directly, not using Mbed OS APIs. It's a bare metal application that we are avoiding modifying for maintainability reasons: it's best if we can avoid any modifications to the application that aren't upstream.

This is the file the application uses for running the timer. The application assumes there is some timer available to NS, available via a HAL-like function. https://git.trustedfirmware.org/trusted-firmware-m.git/tree/platform/ext/target/cypress/psoc64/plat_test.c#n55

morser499 replied:
I understand not wanting to change a code example, but putting this code here increases maintainability overhead. The code would be spread out in two completely different places.

morser499 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_064S2_4343W/device/COMPONENT_CM4/TOOLCHAIN_ARM/cyb06xxa_cm4.sct:
```diff
@@ -27,6 +27,7 @@
 ;*******************************************************************************
 ;* \copyright
 ;* Copyright 2016-2019 Cypress Semiconductor Corporation
+;* Copyright 2020 Arm Limited
```
Why isn't a similar change necessary for the CY8CPROTO_064_SB target?

Patater replied:
What are the long-term plans for `CY8CPROTO_064_SB`? Will Cypress be removing `CY8CPROTO_064_SB` from Mbed OS, along with `CY8CKIT_064S2_4343W`, when final hardware is available?

jainvikas8 replied:
CY8CPROTO_064_SB is not supported by TF-M (master). So we’ll not be able to support that target for PSA in Mbed OS.

0xc0170 commented on targets/TARGET_Cypress/TARGET_PSOC6/TARGET_CY8CKIT_064S2_4343W/device/COMPONENT_CM4/device_cfg.h:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) 2017-2018 Arm Limited
+ * Copyright (c) 2020, Cypress Semiconductor Corporation. All rights reserved.
+ *
```
SPDX identifiers to new files

Patater reviewed (approved):
LGTM

Let's file an issue to revisit how to share resources with TF-M better.
